File: tooling/gen_readme.ts
import { existsSync, readdirSync, readFileSync, statSync, writeFileSync } from "node:fs";
import { join } from "node:path";
import { parseArgs } from "node:util";
import { exitWith, isMain } from "./lib/cli.ts";
import { parseFrontmatter, stringMap } from "./lib/frontmatter.ts";
import { REPO_ROOT } from "./lib/repo.ts";

interface ToolchainSnapshot {
  verification_date: string;
  platform: { os: string; arch: string };
  components: Array<{ name: string; version: string }>;
  verified_targets: string[];
}

interface ActivationPrompt {
  category: string;
}

export function readSnapshot(repoRoot = REPO_ROOT): ToolchainSnapshot {
  return JSON.parse(
    readFileSync(join(repoRoot, "verification", "toolchains", "current.json"), "utf8"),
  ) as ToolchainSnapshot;
}

export function countActivationPrompts(repoRoot = REPO_ROOT): Map<string, number> {
  const counts = new Map<string, number>();
  const prompts = join(repoRoot, "evals", "activation", "prompts.jsonl");
  if (!isFile(prompts)) {
    return counts;
  }
  for (const line of readFileSync(prompts, "utf8").split("\n")) {
    if (line.trim() === "") {
      continue;
    }
    const prompt = JSON.parse(line) as ActivationPrompt;
    counts.set(prompt.category, (counts.get(prompt.category) ?? 0) + 1);
  }
  return counts;
}

export function blockStatus(repoRoot = REPO_ROOT): string {
  const snapshot = readSnapshot(repoRoot);
  const components = new Map(
    snapshot.components.map((component) => [component.name, component.version]),
  );
  return [
    `- Verified toolchain: \`moon ${components.get("moon")}\` · \`moonc v${components.get("moonc")}\` · \`moonrun ${components.get("moonrun")}\``,
    `- Verification date: ${snapshot.verification_date} on ${snapshot.platform.os} ${snapshot.platform.arch}`,
    `- Verified targets: ${snapshot.verified_targets.join(", ")}`,
  ].join("\n");
}

export function blockInventory(repoRoot = REPO_ROOT): string {
  const fixtures = directories(join(repoRoot, "verification", "fixtures")).filter((directory) =>
    isFile(join(directory, "fixture.json")),
  );
  const lines: string[] = [];
  for (const skillDirectory of directories(join(repoRoot, "skills"))) {
    const skillMd = join(skillDirectory, "SKILL.md");
    if (!isFile(skillMd)) {
      continue;
    }
    const parsed = parseFrontmatter(readFileSync(skillMd, "utf8"));
    const referencesDirectory = join(skillDirectory, "references");
    const references = isDirectory(referencesDirectory)
      ? readdirSync(referencesDirectory, { withFileTypes: true })
          .filter((entry) => entry.isFile() && !entry.name.startsWith("."))
          .map((entry) => entry.name)
          .sort()
      : [];
    const bodyLines = parsed.body.split("\n").length;
    const skillName = skillDirectory.split(/[\\/]/).at(-1);
    const version = stringMap(parsed.frontmatter, "metadata")?.["skill-version"] ?? "?";
    lines.push(
      `- \`${skillName}\` v${version}: SKILL.md (${bodyLines} lines) + ${references.length} reference file(s)`,
    );
  }
  lines.push(`- Verification fixtures: ${fixtures.length}`);

  const counts = countActivationPrompts(repoRoot);
  if (counts.size > 0) {
    const sortedCounts = [...counts].sort(([left], [right]) => left.localeCompare(right));
    const total = sortedCounts.reduce((sum, [, count]) => sum + count, 0);
    const perCategory = sortedCounts.map(([category, count]) => `${category} ${count}`).join(", ");
    lines.push(`- Activation eval prompts: ${total} (${perCategory})`);
  }
  return lines.join("\n");
}

const BLOCKS: Record<string, (repoRoot: string) => string> = {
  status: blockStatus,
  inventory: blockInventory,
};

export function generateReadme(text: string, repoRoot = REPO_ROOT): string {
  let generated = text;
  for (const [name, producer] of Object.entries(BLOCKS)) {
    const begin = `<!-- BEGIN GENERATED: ${name} -->`;
    const end = `<!-- END GENERATED: ${name} -->`;
    const pattern = new RegExp(`(${escapeRegExp(begin)}\\n)[\\s\\S]*?(${escapeRegExp(end)})`);
    if (!pattern.test(generated)) {
      throw new Error(`README.md missing generated block markers for '${name}'`);
    }
    generated = generated.replace(
      pattern,
      (_, opening: string, closing: string) => `${opening}${producer(repoRoot)}\n${closing}`,
    );
  }
  return generated;
}

export function main(args = process.argv.slice(2), repoRoot = REPO_ROOT): number {
  const { values } = parseArgs({
    args,
    options: { check: { type: "boolean", default: false } },
    strict: true,
  });
  const readme = join(repoRoot, "README.md");
  const text = readFileSync(readme, "utf8");
  let generated: string;
  try {
    generated = generateReadme(text, repoRoot);
  } catch (error) {
    console.error(`FAIL ${(error as Error).message}`);
    return 1;
  }

  if (generated !== text) {
    if (values.check) {
      console.error("FAIL README.md generated blocks are stale; run vp run gen-readme");
      return 1;
    }
    writeFileSync(readme, generated, "utf8");
    console.log("README.md regenerated");
  } else {
    console.log("README.md generated blocks: up to date");
  }
  return 0;
}

function directories(parent: string): string[] {
  return readdirSync(parent, { withFileTypes: true })
    .filter((entry) => entry.isDirectory())
    .map((entry) => join(parent, entry.name))
    .sort();
}

function isFile(path: string): boolean {
  return existsSync(path) && statSync(path).isFile();
}

function isDirectory(path: string): boolean {
  return existsSync(path) && statSync(path).isDirectory();
}

function escapeRegExp(value: string): string {
  return value.replace(/[.*+?^${}()|[\]\\]/g, "\\$&");
}

if (isMain(import.meta.url)) {
  exitWith(main());
}

File: tooling/gen_readme.py
#!/usr/bin/env python3
"""Regenerate the data-derived blocks in README.md from repository state.

Every number, version, and inventory line in README.md that could drift
lives between markers:

    <!-- BEGIN GENERATED: <block> -->
    ...
    <!-- END GENERATED: <block> -->

This script rewrites those blocks deterministically from the toolchain
snapshot, the fixtures, the eval prompt sets, and the skill directories.
CI reruns it and fails on `git diff --exit-code`, so the README can never
disagree with the repository.

Usage: python3 tooling/gen_readme.py [--check]
"""

from __future__ import annotations

import argparse
import json
import re
import sys
from pathlib import Path

from validate_skills import parse_frontmatter

REPO_ROOT = Path(__file__).resolve().parent.parent
README = REPO_ROOT / "README.md"


def snapshot() -> dict:
    return json.loads(
        (REPO_ROOT / "verification" / "toolchains" / "current.json").read_text()
    )


def count_activation_prompts() -> dict[str, int]:
    counts: dict[str, int] = {}
    prompts = REPO_ROOT / "evals" / "activation" / "prompts.jsonl"
    if prompts.is_file():
        for line in prompts.read_text().splitlines():
            if not line.strip():
                continue
            counts[json.loads(line)["category"]] = (
                counts.get(json.loads(line)["category"], 0) + 1
            )
    return counts


def block_status() -> str:
    snap = snapshot()
    comps = {c["name"]: c["version"] for c in snap["components"]}
    lines = [
        f"- Verified toolchain: `moon {comps['moon']}` · `moonc v{comps['moonc']}` · "
        f"`moonrun {comps['moonrun']}`",
        f"- Verification date: {snap['verification_date']} on "
        f"{snap['platform']['os']} {snap['platform']['arch']}",
        f"- Verified targets: {', '.join(snap['verified_targets'])}",
    ]
    return "\n".join(lines)


def block_inventory() -> str:
    fixtures = sorted(
        d.name
        for d in (REPO_ROOT / "verification" / "fixtures").iterdir()
        if (d / "fixture.json").is_file()
    )
    lines = []
    for skill_dir in sorted((REPO_ROOT / "skills").iterdir()):
        if not (skill_dir / "SKILL.md").is_file():
            continue
        fm, body, _ = parse_frontmatter((skill_dir / "SKILL.md").read_text())
        refs = sorted(
            f.name
            for f in (skill_dir / "references").glob("*")
            if f.is_file() and not f.name.startswith(".")
        )
        body_lines = body.count("\n") + 1
        lines.append(
            f"- `{skill_dir.name}` v{fm.get('metadata', {}).get('skill-version', '?')}: "
            f"SKILL.md ({body_lines} lines) + {len(refs)} reference file(s)"
        )
    lines.append(f"- Verification fixtures: {len(fixtures)}")
    counts = count_activation_prompts()
    if counts:
        total = sum(counts.values())
        per = ", ".join(f"{k} {v}" for k, v in sorted(counts.items()))
        lines.append(f"- Activation eval prompts: {total} ({per})")
    return "\n".join(lines)


BLOCKS = {
    "status": block_status,
    "inventory": block_inventory,
}


def main() -> int:
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument("--check", action="store_true", help="fail if README is stale")
    args = parser.parse_args()

    text = README.read_text()
    new_text = text
    for name, producer in BLOCKS.items():
        pattern = re.compile(
            rf"(<!-- BEGIN GENERATED: {name} -->\n)(?:.*\n)??(<!-- END GENERATED: {name} -->)",
            re.S,
        )
        if not pattern.search(new_text):
            print(f"FAIL README.md missing generated block markers for {name!r}", file=sys.stderr)
            return 1
        new_text = pattern.sub(
            lambda m: m.group(1) + producer() + "\n" + m.group(2), new_text
        )

    if new_text != text:
        if args.check:
            print("FAIL README.md generated blocks are stale; run tooling/gen_readme.py", file=sys.stderr)
            return 1
        README.write_text(new_text)
        print("README.md regenerated")
    else:
        print("README.md generated blocks: up to date")
    return 0


if __name__ == "__main__":
    sys.exit(main())
